README.md
![aglio](https://raw.github.com/as-op/aglio/main/images/aglio.png)


[![license](https://img.shields.io/github/license/as-op/aglio.svg)](http://opensource.org/licenses/MIT)

# Introduction

An [API Blueprint](http://apiblueprint.org/) renderer that supports multiple themes and outputs static HTML that can be served by any web host. API Blueprint is a Markdown-based document format that lets you write API descriptions and documentation in a simple and straightforward way. Currently supported is [API Blueprint format 1A](https://github.com/apiaryio/api-blueprint/blob/master/API%20Blueprint%20Specification.md).

Note: **This is an experiment to upgrade** in forking [aglio](https://github.com/danielgtaylor/aglio), which is unmaintained.

## Features

 * Parsing with [drafter.js](https://github.com/apiaryio/drafter.js/)
 * Asyncronous processing
 * Multiple templates/themes
 * Support for custom colors, templates, and theme engines
 * Include other documents in your blueprint
 * Commandline executable `aglio -i service.apib -o api.html`
 * Live-reloading preview server `aglio -i service.apib --server`
 * Node.js library `require('aglio')`
 * Excellent test coverage
 * ESLinted

## Example Output
Example output is generated from the [example API Blueprint](https://raw.github.com/as-op/aglio/main/example.apib) using the default [Olio theme](https://github.com/as-op/aglio/tree/main/lib/aglio-theme-olio).

 * Default theme [two column](https://htmlpreview.github.io/?https://raw.githubusercontent.com/as-op/aglio/blob/main/examples/default.html) or [three column](https://htmlpreview.github.io/?https://raw.githubusercontent.com/as-op/aglio/blob/main/examples/default-triple.html)
 * Streak theme [two column](https://htmlpreview.github.io/?https://raw.githubusercontent.com/as-op/aglio/blob/main/examples/streak.html) or [three column](https://htmlpreview.github.io/?https://raw.githubusercontent.com/as-op/aglio/blob/main/examples/streak-triple.html)
 * Flatly theme [two column](https://htmlpreview.github.io/?https://raw.githubusercontent.com/as-op/aglio/blob/main/examples/flatly.html) or [three column](https://htmlpreview.github.io/?https://raw.githubusercontent.com/as-op/aglio/blob/main/examples/flatly-triple.html)
 * Slate theme [two column](https://htmlpreview.github.io/?https://raw.githubusercontent.com/as-op/aglio/blob/main/examples/slate.html) or [three column](https://htmlpreview.github.io/?https://raw.githubusercontent.com/as-op/aglio/blob/main/examples/slate-triple.html)
 * Cyborg theme [two column](https://htmlpreview.github.io/?https://raw.githubusercontent.com/as-op/aglio/blob/main/examples/cyborg.html) or [three column](https://htmlpreview.github.io/?https://raw.githubusercontent.com/as-op/aglio/blob/main/examples/cyborg-triple.html)

## Including Files
It is possible to include other files in your blueprint by using a special include directive with a path to the included file relative to the current file's directory. Included files can be written in API Blueprint, Markdown or HTML (or JSON for response examples). Included files can include other files, so be careful of circular references.

```markdown
<!-- include(filename.md) -->
```

For tools that do not support this include directive it will just render out as an HTML comment. API Blueprint may support its own mechanism of including files in the future, and this syntax was chosen to not interfere with the [external documents proposal](https://github.com/apiaryio/api-blueprint/issues/20) while allowing `aglio` users to include documents today.


# Installation & Usage
There are two ways to use aglio: as an executable or as a library for Node.js.

## Executable
Install aglio via NPM. You need Node.js installed.

```bash
npm install -g aglio
```

Then, start generating HTML.

```bash
# Default theme
aglio -i input.apib -o output.html

# Use three-column layout
aglio -i input.apib --theme-template triple -o output.html

# Built-in color scheme
aglio --theme-variables slate -i input.apib -o output.html

# Customize a built-in style
aglio --theme-style default --theme-style ./my-style.less -i input.apib -o output.html

# Custom layout template
aglio --theme-template /path/to/template.pug -i input.apib -o output.html

# Custom theme engine
aglio -t my-engine -i input.apib -o output.html

# Run a live preview server on http://localhost:3000/
aglio -i input.apib -s

# Print output to terminal (useful for piping)
aglio -i input.apib -o -

# Disable condensing navigation links
aglio --no-theme-condense -i input.apib -o output.html

# Render full-width page instead of fixed max width
aglio --theme-full-width -i input.apib -o output.html

# Set an explicit file include path and read from stdin
aglio --include-path /path/to/includes -i - -o output.html

# Output verbose error information with stack traces
aglio -i input.apib -o output.html --verbose
```

## Node.js Library
You can also use aglio as a library. First, install and save it as a dependency:

```bash
npm install --save aglio
```

Then, convert some API Blueprint to HTML:

```javascript
var aglio = require('aglio');

// Render a blueprint with a template by name
var blueprint = '# Some API Blueprint string';
var options = {
  themeVariables: 'default'
};

aglio.render(blueprint, options, function (err, html, warnings) {
    if (err) return console.log(err);
    if (warnings) console.log(warnings);

    console.log(html);
});

// Render a blueprint with a custom template file
options = {
  themeTemplate: '/path/to/my-template.pug'
};
aglio.render(blueprint, options, function (err, html, warnings) {
    if (err) return console.log(err);
    if (warnings) console.log(warnings);

    console.log(html);
});


// Pass custom locals along to the template, for example
// the following gives templates access to lodash and async
options = {
    themeTemplate: '/path/to/my-template.pug',
    locals: {
        _: require('lodash'),
        async: require('async')
    }
};
aglio.render(blueprint, options, function (err, html, warnings) {
   if (err) return console.log(err);
   if (warnings) console.log(warnings);

   console.log(html);
});
```

### Reference
The following methods are available from the `aglio` library:

#### aglio.collectPathsSync (blueprint, includePath)
Get a list of paths that are included in the blueprint. This list can be watched for changes to do things like live reload. The blueprint's own path is not included.

```javascript
var blueprint = '# GET /foo\n<-- include(example.json -->\n';
var watchPaths = aglio.collectPathsSync(blueprint, process.cwd())
```

#### aglio.render (blueprint, options, callback)
Render an API Blueprint string and pass the generated HTML to the callback. The `options` can either be an object of options or a simple layout name or file path string. Available options are:

| Option      | Type   | Default       | Description                           |
| ----------- | ------ | ------------- | ------------------------------------- |
| filterInput | bool   | `true`        | Filter `\r` and `\t` from the input   |
| includePath | string | process.cwd() | Base directory for relative includes  |
| locals      | object | `{}`          | Extra locals to pass to templates     |
| theme       | string | `'default'`   | Theme name to load for rendering      |

In addition, the [default theme](https://github.com/as-op/aglio/tree/main/lib/aglio-theme-olio) provides the following options:

| Option                | Type   | Default   | Description                                  |
| --------------------- | ------ | --------- | -------------------------------------------- |
| themeVariables        | string | `default` | Built-in color scheme or path to LESS or CSS |
| themeCondenseNav      | bool   | `true`    | Condense single-action navigation links      |
| themeFullWidth        | bool   | `false`   | Use the full page width                      |
| themeTemplate         | string |           | Layout name or path to custom layout file    |
| themeStyle            | string | `default` | Built-in style name or path to LESS or CSS   |
| themePermalinkSymbol  | string | `¶`       | The symbol in the permalink anchor.          |


```javascript
var blueprint = '...';
var options = {
    themeTemplate: 'default',
    locals: {
        myVariable: 125
    }
};

aglio.render(blueprint, options, function (err, html, warnings) {
    if (err) return console.log(err);

    console.log(html);
});
```

#### aglio.renderFile (inputFile, outputFile, options, callback)
Render an API Blueprint file and save the HTML to another file. The input/output file arguments are file paths. The options behaves the same as above for `aglio.render`, except that the `options.includePath` defaults to the basename of the input filename.

```javascript
aglio.renderFile('/tmp/input.apib', '/tmp/output.html', options, function (err, warnings) {
    if (err) return console.log(err);
    if (warnings) console.log(warnings);
});
```

# Development
Pull requests are encouraged! Feel free to fork and hack away, especially on new themes. The build system in use is Grunt, so make sure you have it installed:

```bash
# Lint/compile the js
npm run lint

# Run the test suite
npm run test

# Generate an HTML test coverage report
npm run coverage

# Render examples
npm run gen:examples
```

## Customizing Output
Aglio is split into two components: a base that contains logic for loading API Blueprint, handling commandline arguments, etc and a theme engine that handles turning the API Blueprint AST into HTML. The default theme engine that ships with Aglio is called olio, included in this repository. Templates are written in pug. The default stylesheets are written in LESS.

While developing customizations, you may want to disable caching using the `NOCACHE` environment variable.

```bash
NOCACHE=1 aglio -i input.apib [customization options]
```

### Custom Colors & Style
Aglio's default theme provides a way to easily override colors, fonts, padding, etc to match your company's style. This is done by providing your own LESS or CSS file(s) via the `--theme-variables` and `--theme-style` options. For example:

```bash
# Use my custom colors
aglio --theme-variables /path/to/my-colors.less -i input.apib -o output.html
```

The `my-variables.less` file might contain a custom HTTP PUT color specification:

```less
/* HTTP PUT */
@put-color: #f0ad4e;
@put-background-color: #fcf8e3;
@put-text-color: contrast(@put-background-color);
@put-border-color: darken(spin(@put-background-color, -10), 5%);
```

See the [default variables](https://github.com/as-op/aglio/blob/main/lib/aglio-theme-olio/styles/variables-default.less) file for examples of which variables can be set.

The `--theme-style` option lets you override built-in styles with your own LESS or CSS definitions. It is processed **after** the variables have been defined, so the variables are available for your use. If you wish to modify a rule from an existing built-in style then you must copy the style. The order of loading roughly follows:

1. Default variables
2. Built-in **or** user-supplied variables
3. Built-in **or** user-supplied style

Note that these options can be passed more than once, in which case they will be loaded in the order they were passed. This lets you, for example, load a variable preset like `flatly` and modify one of the colors with your own LESS file. Keep in mind that when you want to modify a built-in style you must explicitly list the style, e.g. `--theme-style default --theme-style my-style.less`.

#### Built-in Colors

* `cyborg`
* `default`
* `flatly`
* `slate`

#### Built-in Styles

* `default`

### Customizing Layout Templates
The `--theme-template` option allows you to provide a custom layout template that overrides the default. This is specified in the form of a [Pug](https://pugjs.org) template file. See the [default template](https://github.com/as-op/aglio/blob/main/lib/aglio-theme-olio/templates/index.pug) file for an example.

The locals available to templates look like the following:

| Name        | Description                                              |
| ----------- | -------------------------------------------------------- |
| api         | The API Blueprint AST from Protagonist                   |
| condenseNav | If true, you should condense the nav if possible         |
| date        | Date and time handling from Moment.js                    |
| fullWidth   | If true, you should consume the entire page width        |
| highlight   | A function (`code`, `lang`) to highlight a piece of code |
| markdown    | A function to convert Markdown strings to HTML           |
| slug        | A function to convert a string to a slug usable as an ID |
| hash        | A function to return an hash (currently MD5)             |

#### Built-in Layout Templates

* `default`

### Using Custom Themes
While Aglio ships with a default theme, you have the option of installing and using third-party theme engines. They may use any technology and are not limited to Pug and LESS. Consult the theme's documentation to see which options are available and how to use and customize the theme. Common usage between all themes:

```bash
# Install a custom theme engine globally
npm install -g aglio-theme-<NAME>

# Render using a custom theme engine
aglio -t <NAME> -i input.apib -o output.html

# Get a list of all options for a theme
aglio -t <NAME> --help
```

### Writing a Theme Engine
Theme engines are simply Node.js modules that provide two public functions and follow a specific naming scheme (`aglio-theme-NAME`). Because they are their own npm package they can use whatever technologies the theme engine author wishes. The only hard requirement is to provide these two public functions:

#### `getConfig()`
Returns configuration information about the theme, such as the API Blueprint format that is supported and any options the theme provides.

#### `render(input, options, done)`
Render the given input API Blueprint AST with the given options. Calls `done(err, html)` when finished, either passing an error or the rendered HTML output as a string.

#### Example Theme
The following is a very simple example theme. **Note**: it only returns a very simple string instead of rending out the API Blueprint AST. Normally you would invoke a template engine and output the resulting HTML that is generated.

```javascript
// Get the theme's configuration options
exports.getConfig = function () {
  return {
    // This is a list of all supported API Blueprint format versions
    formats: ['1A'],
    // This is a list of all options your theme accepts. See
    // here for more: https://github.com/bcoe/yargs#readme
    // Note: These get prefixed with `theme` when you access
    // them in the options object later!
    options: [
      {
        name: 'name',
        description: 'Your name',
        default: 'world'
      }
    ]
  };
}

// Asyncronously render out a string
exports.render = function (input, options, done) {
  // Normally you would use some template engine here.
  // To keep this code really simple, we just print
  // out a string and ignore the API Blueprint.
  done(null, 'Hello, ' + options.themeName + '!');
};

```

Example use:

```bash
# Install the theme globally
npm install -g aglio-theme-hello

# Render some output!
aglio -t hello -i example.apib -o -
=> 'Hello, world!'

# Pass in the custom theme option!
aglio -t hello --theme-name Daniel -i example.apib -o -
=> 'Hello, Daniel!'
```

You are free to use whatever template system (Pug, EJS, Nunjucks, etc) and any supporting libraries (e.g. for CSS) you like.

License
=======
Copyright (c) 2016 Daniel G. Taylor

http://dgt.mit-license.org/


CHANGELOG.md
# fork

* Switch to [drafter.js](https://github.com/apiaryio/drafter.js/) to avoid native compilation problems
* Include default theme "olio", it was integrated and required anyhow
* Get rid of Coffee-Script, use pure ES
* Update dependencies & adapt to them
* ESLint
* Cherry-picked changes by https://github.com/martide/aglio

# 2.3.0

* Switch to [Drafter](https://github.com/apiaryio/drafter-npm) for API Blueprint parsing. [#277](https://github.com/danielgtaylor/aglio/pull/277)
* Fix case when no arguments are supplied after `-o` [#262](https://github.com/danielgtaylor/aglio/pull/262)

# 2.2.1 - 2016-05-20

* Bump protagonist version to 1.3.2 to support new features. [#256](https://github.com/danielgtaylor/aglio/pull/256)
* Update required default theme version to [1.6.3](https://github.com/danielgtaylor/aglio/blob/olio-theme/Changelog.md#163---2016-05-20).

# 2.2.0 - 2015-10-26

* Support for Node.js 3.x and 4.x. [#183](https://github.com/danielgtaylor/aglio/issues/183)
* Upgraded to [Protagonist](https://github.com/apiaryio/protagonist) version 1.x, which has disabled the JSON Schema generation from MSON until some bugs can be worked out.

# 2.1.1 - 2015-09-14

* Fix the default include path behavior when outputting a file to use the basename of the input file rather than `process.cwd()`. Also updates documentation to mention this behavior and option. [#166](https://github.com/danielgtaylor/aglio/issues/166)

# 2.1.0 - 2015-09-11

* Add a `--include-path` option to set the path for relative includes. [#165](https://github.com/danielgtaylor/aglio/pull/165)

# 2.0.4 - 2015-08-14

* Show theme name before loading when given `--verbose` option.
* Update required default theme version to [1.2.0](https://github.com/danielgtaylor/aglio/blob/olio-theme/Changelog.md#120---2015-08-14).

# 2.0.3 - 2015-08-12

* Update required default theme version to [1.1.0](https://github.com/danielgtaylor/aglio/blob/olio-theme/Changelog.md#110---2015-08-12).
* Readme and example updates.

# 2.0.2 - 2015-08-04

* Update required default theme version to support custom jade templates written for Aglio 1.x.
* Add a `--verbose` output option for errors that displays pretty stack traces.
* Provide more descriptive errors.

# 2.0.1 - 2015-08-03

* Add backward-compatible support for the `aglio` binary to specify a custom layout via the `-t` commandline option. Example: `aglio -i input.apib -t /path/to/my.jade -o output.html`
* Display default theme version when using `aglio --version`.

# 2.0.0 - 2015-07-16
This is a new major version of Aglio, and as such has some breaking changes.
High-level changes in this release:

* Use [Drafter.js](https://github.com/apiaryio/drafter.js) to support [MSON](https://github.com/apiaryio/mson) via generated request/response bodies and schemas.
* Add support for [theme engines](https://github.com/danielgtaylor/aglio#using-custom-themes).
* Use [Olio](https://github.com/danielgtaylor/aglio/tree/olio-theme#readme) as the new default theme engine.
* API and resource group description headers are now included in the navigation bar.
* Server mode can now serve static files, which is useful if your documentation contains images.
* Fixes to how resource and action parameters are handled.

For more detailed information, see the beta releases below.

Effort was taken to prevent backward-incompatible changes. Here is a list of
things that **will break** if you used them in 1.x.

Binary:
* It is no longer possible to list templates (`aglio -l`). You may use `npm list -g | grep aglio-theme` to list all installed theme engine packages instead. Refer to individual theme documentation for possible theme engine options.

Library:
* The `aglio.getTemplates` function has been **removed**.

Templates:
* The multi-page layouts have been **removed**. Please open an issue if you would like to see them in the new theme engine.
* The collapsed navigation layouts have been **removed**. This is now in the default theme and handled automatically based on browser window height.

The following are translated internally and will not break, but are suggested updates:

Binary:
* The `-t` option is now shorthand for `--theme` instead of `--template`.
* The `--full-width` and `--condense-nav` parameters are now `--theme-full-width` and `--theme-condense-nav` in the default theme engine.

Library:
* Passing a string as the options to `render` and `renderFile` will still work if it is a known variant: `default`, `flatly`, `slate`, `cyborg` or one of the collapsed versions of those. If it is a path and the file exists, then it will use it as a custom `themeLayout` option. Otherwise it will set the theme engine name. :dizzy_face:
* You should use `options.theme` instead of `options.template`.
* You should use `options.themeVariables = 'flatly'` to set the color variation.
* You should use `options.themeTemplate = '/path/to/layout.jade'` to set the layout template.

Thank you to all the contributors and testers for helping to make this an awesome release! :beers:

# 2.0.0-beta6 - 2015-07-14
* Update to [olio theme](https://github.com/danielgtaylor/aglio/blob/olio-theme/Changelog.md) 0.0.9.

# 2.0.0-beta5 - 2015-07-10
* Fix an issue with included paths when using `--server`.
* Update to [olio theme](https://github.com/danielgtaylor/aglio/blob/olio-theme/Changelog.md) 0.0.8.

# 2.0.0-beta4 - 2015-07-10
* Update to [olio theme](https://github.com/danielgtaylor/aglio/blob/olio-theme/Changelog.md) 0.0.7.

# 2.0.0-beta3 - 2015-07-09
* Documentation updates.
* Server mode now serves static files if found.
* Add ability to output compiled API Blueprint file instead of HTML.
* Update to [olio theme](https://github.com/danielgtaylor/aglio/blob/olio-theme/Changelog.md) 0.0.6.

# 2.0.0-beta2 - 2015-05-29
* Live update fixes.
* Example fixes.

# 2.0.0-beta1 - 2015-05-28
* Implement theme engine support; depend on the default olio theme.
* Switch to using drafter.js instead of protagonist directly.

# 1.18.0 - 2015-03-31

test/renderer.js
const aglio = require('../lib/main');
const assert = require('assert');
const fs = require('fs');
const path = require('path');
const drafter = require('drafter.js');
const sinon = require('sinon');

const root = path.dirname(__dirname);
const blueprint = fs.readFileSync(path.join(root, 'example.apib'), 'utf-8');

describe('API Blueprint Renderer', function () {
	it('Should load the default theme', function () {
		const theme = aglio.getTheme('default');

		return assert.ok(theme);
	});

	it('Should get a list of included files', function () {
		sinon
			.stub(fs, 'readFileSync')
			.callsFake(() => 'I am a test file');

		const input = `\
# Title
<!-- include(test1.apib) -->
Some content...
<!-- include(test2.apib) -->
More content...\
`;

		const paths = aglio.collectPathsSync(input, '.');

		fs.readFileSync.restore();

		assert.equal(paths.length, 2);
		assert(Array.from(paths).includes('test1.apib'));
		return assert(Array.from(paths).includes('test2.apib'));
	});

	it('Should render blank string', function (done) {
		aglio.render('', {template: 'default', locals: {foo: 1}}, function (err, html) {
			if (err) {
				return done(err);
			}

			assert(html);

			return done();
		});
	});

	it('Should render a complex document', function (done) {
		aglio.render(blueprint, 'default', function (err, html) {
			if (err) {
				return done(err);
			}

			assert(html);

			// Ensure include works
			assert(html.indexOf('This is content that was included'));

			return done();
		});
	});

	it('Should render mixed line endings and tabs properly', function (done) {
		const temp = '# GET /message\r\n+ Response 200 (text/plain)\r\r\t\tHello!\n';
		aglio.render(temp, 'default', done);
	});

	it('Should render a custom template by filename', function (done) {
		const template = path.join(root, 'test', 'test.pug');
		aglio.render('# Blueprint', template, function (err, html) {
			if (err) {
				return done(err);
			}

			assert(html);

			return done();
		});
	});

	it('Should return warnings with filtered input', function (done) {
		const temp = '# GET /message\r\n+ Response 200 (text/plain)\r\r\t\tHello!\n';
		const filteredTemp = temp.replace(/\r\n?/g, '\n').replace(/\t/g, '    ');

		aglio.render(temp, 'default', function (err, html, warnings) {
			if (err) {
				return done(err);
			}

			assert.equal(filteredTemp, warnings.input);

			return done();
		});
	});

	it('Should render from/to files', function (done) {
		const src = path.join(root, 'example.apib');
		const dest = path.join(root, 'example.html');
		aglio.renderFile(src, dest, {}, done);
	});

	it('Should render from stdin', function (done) {
		sinon
			.stub(process.stdin, 'read')
			.callsFake(() => '# Hello\n');

		setTimeout(() => process.stdin.emit('readable', 1));

		aglio.renderFile('-', 'example.html', 'default', function (err) {
			if (err) {
				return done(err);
			}

			assert(process.stdin.read.called);
			process.stdin.read.restore();
			process.stdin.removeAllListeners();

			return done();
		});
	});

	it('Should render to stdout', function (done) {
		sinon.stub(console, 'log');

		aglio.renderFile(path.join(root, 'example.apib'), '-', 'default', function (err) {
			if (err) {
				console.log.restore();
				return done(err);
			}

			assert(console.log.called);
			console.log.restore();

			return done();
		});
	});

	it('Should compile from/to files', function (done) {
		const src = path.join(root, 'example.apib');
		const dest = path.join(root, 'example-compiled.apib');
		aglio.compileFile(src, dest, done);
	});

	it('Should compile from stdin', function (done) {
		sinon
			.stub(process.stdin, 'read')
			.callsFake(() => '# Hello\n');

		setTimeout(() => process.stdin.emit('readable', 1));

		aglio.compileFile('-', 'example-compiled.apib', function (err) {
			if (err) {
				return done(err);
			}

			assert(process.stdin.read.called);
			process.stdin.read.restore();
			process.stdin.removeAllListeners();

			return done();
		});
	});

	it('Should compile to stdout', function (done) {
		sinon.stub(console, 'log');

		aglio.compileFile(path.join(root, 'example.apib'), '-', function (err) {
			if (err) {
				return done(err);
			}

			assert(console.log.called);
			console.log.restore();

			return done();
		});
	});

	it('Should support legacy theme names', function (done) {
		aglio.render('', {template: 'flatly'}, function (err, html) {
			if (err) {
				return done(err);
			}
			assert(html);
			done();
		});
	});

	it('Should error on missing input file', function (done) {
		aglio.renderFile('missing', 'output.html', 'default', function (err, html) {
			assert(err);

			aglio.compileFile('missing', 'output.apib', function (err) {
				assert(err);
				done();
			});
		});
	});

	it('Should error on bad template', function (done) {
		aglio.render(blueprint, 'bad', function (err, html) {
			assert(err);
			done();
		});
	});

	it('Should error on drafter failure', function (done) {
		sinon
			.stub(drafter, 'parse')
			.callsFake((content, options, callback) => callback('error'));

		aglio.render(blueprint, 'default', function (err, html) {
			assert(err);
			drafter.parse.restore();
			done();
		});
	});

	it('Should error on file read failure', function (done) {
		sinon
			.stub(fs, 'readFile')
			.callsFake((filename, options, callback) => callback('error'));

		aglio.renderFile('foo', 'bar', 'default', function (err, html) {
			assert(err);
			fs.readFile.restore();
			done();
		});
	});

	it('Should error on file write failure', function (done) {
		sinon
			.stub(fs, 'writeFile')
			.callsFake((filename, data, callback) => callback('error'));
		aglio.renderFile('foo', 'bar', 'default', function (err, html) {
			assert(err);
			fs.writeFile.restore();
			done();
		});
	});

	return it('Should error on non-file failure', function (done) {
		sinon
			.stub(aglio, 'render')
			.callsFake((content, template, callback) => callback('error'));

		aglio.renderFile(path.join(root, 'example.apib'), 'bar', 'default', function (err, html) {
			assert(err);

			aglio.render.restore();

			return done();
		});
	});
});


lib/bin.js
const aglio = require('./main');
const chokidar = require('chokidar');
const clc = require('cli-color');
const fs = require('fs');
const http = require('http');
const path = require('path');
const PrettyError = require('pretty-error');
const serveStatic = require('serve-static');
const parser = require('yargs')
	.usage('Usage: $0 [options] -i infile [-o outfile -s]')
	.example('$0 -i example.apib -o output.html', 'Render to HTML')
	.example('$0 -i example.apib -s', 'Start preview server')
	.example('$0 --theme-variables flatly -i example.apib -s', 'Theme colors')
	.example('$0 --no-theme-condense-nav -i example.apib -s', 'Disable options')
	.options('i', { alias: 'input', describe: 'Input file' })
	.options('o', { alias: 'output', describe: 'Output file' })
	.options('t', { alias: 'theme', describe: 'Theme name or layout file', default: 'default' })
	.options('f', { alias: 'filter', boolean: true, describe: 'Sanitize input from Windows', default: true })
	.options('s', { alias: 'server', describe: 'Start a local live preview server' })
	.options('h', { alias: 'host', describe: 'Address to bind local preview server to', default: '127.0.0.1' })
	.options('p', { alias: 'port', describe: 'Port for local preview server', default: 3000 })
	.options('v', { alias: 'version', describe: 'Display version number', default: false })
	.options('c', { alias: 'compile', describe: 'Compile the blueprint file', default: false })
	.options('n', { alias: 'include-path', describe: 'Base directory for relative includes' })
	.options('verbose', { describe: 'Show verbose information and stack traces', default: false })
	.epilog('See https://github.com/danielgtaylor/aglio#readme for more information');

// Console color settings for error/warnings
const cErr = clc.white.bgRed;
const cWarn = clc.xterm(214).bgXterm(235);

// Get the context from an error if possible
const getErrContext = function (input, lineNo) {
	const inputLines = input.split('\n');
	const context = inputLines.slice(lineNo - 5, lineNo + 5);
	return context.map(function (line, index) {
		if (index === 4) {
			return cWarn(`>>>>   ${line}`);
		} else {
			return `       ${line}`;
		}
	});
};

// Get a line number from an error if possible
const getLineNo = function (input, err) {
	if (err.location && err.location.length) {
		return input.substr(0, err.location[0].index).split('\n').length;
	}
};

// Output warning info
const logWarnings = (warnings) => {
	const result = [];
	for (const warning of Array.from(warnings || [])) {
		const lineNo = getLineNo(warnings.input, warning) || 0;
		const errContext = getErrContext(warnings.input, lineNo);
		console.error(cWarn(`>> Line ${lineNo}:`) + ` ${warning.message} (warning code ${warning.code})`);
		console.error(cWarn('>> Context'));
		const err = `       ...\n ${errContext.join('\n')} \n       ...`;
		console.error(err);
		result.push(err);
	}
	return result;
};

// Output an error message
const logError = function (err, verbose) {
	if (verbose) {
		const pe = new PrettyError();
		pe.setMaxItems(5);
		return console.error(pe.render(err));
	} else {
		return console.error(cErr('>>'), err);
	}
};

exports.live = {};

exports.run = function (argv, done) {
	let theme;
	if (argv == null) {
		({ argv } = parser);
	}
	if (done == null) {
		done = function () {
		};
	}
	let _html = null;
	const getHtml = function (cb) {
		if (_html) {
			return cb && cb(null, _html);
		} else {
			return fs.readFile(argv.i, 'utf-8', function (err, blueprint) {
				if (err) {
					logError(err, argv.verbose);
					return cb && cb(err);
				}
				console.log('Rendering ' + argv.i);
				return aglio.render(blueprint, argv, function (err, html, warnings) {
					logWarnings(warnings);
					if (err) {
						logError(err, argv.verbose);
						return cb && cb(err);
					} else {
						_html = html;
						return cb && cb(null, _html);
					}
				});
			});
		}
	};

	if (argv.version) {
		console.log(`aglio ${require('../package.json').version}`);
		return done();
	}

	// The option used to be called `template`
	if (argv.template) {
		argv.theme = argv.template;
	}

	// Backward-compatible support for -t /path/to/layout.jade
	if (fs.existsSync(argv.theme)) {
		argv.themeTemplate = argv.theme;
		argv.theme = 'default';
	}

	// Add theme options to the help output
	if (argv.verbose) {
		console.log(`Loading theme ${argv.theme}`);
	}
	try {
		theme = aglio.getTheme(argv.theme);
	} catch (error) {
		const err = error;
		err.message = `Could not load theme: ${err.message}`;
		logError(err, argv.verbose);
		return done(err);
	}

	const config = theme.getConfig();
	for (const entry of Array.from(config.options)) {
		parser.options(`theme-${entry.name}`, entry);
	}

	if (argv.s) {
		if (!argv.i) {
			parser.showHelp();
			return done('Invalid arguments');
		}

		argv.locals =
				{ livePreview: true };

		// Set where to include files from before generating HTML
		if (argv.i !== '-') {
			argv.includePath = path.dirname(argv.i);
		}

		getHtml();
		const server = http.createServer(function (req, res) {
			if (req.url !== '/') {
				const serve = serveStatic(path.dirname(argv.i));
				return serve(req, res, () => res.end());
			}

			return getHtml(function (err, html) {
				res.writeHead(200, { 'Content-Type': 'text/html' });
				return res.end((err ? err.toString() : html));
			});
		}).listen(argv.p, argv.h, () => console.log(`Server started on http://${argv.h}:${argv.p}/`));

		const sendHtml = socket => getHtml(function (err, html) {
			if (!err) {
				console.log('Refresh web page in browser');
				const re = /<body.*?>[^]*<\/body>/gi;
				html = html.match(re)[0];
				return socket.emit('refresh', html);
			}
		});

		const io = require('socket.io')(server);
		io.on('connection', function (socket) {
			console.log('Socket connected');
			return socket.on('request-refresh', () => sendHtml(socket));
		});

		const paths = aglio.collectPathsSync(fs.readFileSync(argv.i, 'utf-8'), path.dirname(argv.i));

		const watcher = chokidar.watch([argv.i].concat(paths));
		watcher.on('change', function (path) {
			console.log('Updated ' + path);
			_html = null;
			return sendHtml(io);
		});
		exports.live = { server, io, watcher };
		return done();
	} else {
		// Render or Compile API Blueprint, requires input/output files
		if (!argv.i || !argv.o) {
			parser.showHelp();
			return done('Invalid arguments');
		}

		if (argv.c || ((typeof argv.o === 'string') && (argv.o.match(/\.apib$/ || argv.o.match(/\.md$/))))) {
			return aglio.compileFile(argv.i, argv.o, function (err) {
				if (err) {
					logError(err, argv.verbose);
				}

				return done();
			});
		} else {
			return aglio.renderFile(argv.i, argv.o, argv, function (err, warnings) {
				if (err) {
					const lineNo = getLineNo(err.input, err);
					if (lineNo != null) {
						console.error(cErr(`>> Line ${lineNo}:`) + ` ${err.message} (error code ${err.code})`);
					} else {
						logError(err, argv.verbose);
					}

					return done(err);
				}

				logWarnings(warnings);

				return done();
			});
		}
	}
};
